### ISSUE TITLE
'Y' and 'R' button not detected by SDL applications

### ISSUE THREAD

**Issue by terencode** [atar-axis/xpadneo#228]
## Version of xpadneo
v0.8.r515.g715c2e9

## Describe the bug
When connecting an 8bitdo controller (tested with both the zero 2 and SN30 pro+) in X Input Mode over bluetooth,
the keys labeled 'Y' and 'R' are not registered in sdl controllermap and antimicrox.

## Steps to Reproduce

1. Connect an 8bitdo zero 2 controller over bluetooth in X Input Mode (Start + X)
2. Run jstest/jstest-gtk and notice all buttons are correctly registered
3. Run antimicrox or sdl controllermap and notice the 'Y' and 'R' buttons do not do anything

## Expected behavior

All the buttons should be detected.

## System information

```console
# uname -a
Linux terence-desktop 5.7.8-18-tkg-bmq #1 TKG SMP PREEMPT Fri, 10 Jul 2020 18:00:53 +0000 x86_64 GNU/Linux
```

```console
# xxd -c20 -g1 /sys/module/hid_xpadneo/drivers/hid:xpadneo/0005:045E:*/report_descriptor | tee >(cksum)
00000000: 05 01 09 05 a1 01 85 01 09 01 a1 00 09 30 09 31 15 00 27 ff  .............0.1..'.
00000014: ff 00 00 95 02 75 10 81 02 c0 09 01 a1 00 09 33 09 34 15 00  .....u.........3.4..
00000028: 27 ff ff 00 00 95 02 75 10 81 02 c0 05 01 09 32 15 00 26 ff  '......u.......2..&.
0000003c: 03 95 01 75 0a 81 02 15 00 25 00 75 06 95 01 81 03 05 01 09  ...u.....%.u........
00000050: 35 15 00 26 ff 03 95 01 75 0a 81 02 15 00 25 00 75 06 95 01  5..&....u.....%.u...
00000064: 81 03 05 01 09 39 15 01 25 08 35 00 46 3b 01 66 14 00 75 04  .....9..%.5.F;.f..u.
00000078: 95 01 81 42 75 04 95 01 15 00 25 00 35 00 45 00 65 00 81 03  ...Bu.....%.5.E.e...
0000008c: 05 09 19 01 29 0a 15 00 25 01 75 01 95 0a 81 02 15 00 25 00  ....)...%.u.......%.
000000a0: 75 06 95 01 81 03 05 01 09 80 85 02 a1 00 09 85 15 00 25 01  u.................%.
000000b4: 95 01 75 01 81 02 15 00 25 00 75 07 95 01 81 03 c0 05 0f 09  ..u.....%.u.........
000000c8: 21 85 03 a1 02 09 97 15 00 25 01 75 04 95 01 91 02 15 00 25  !........%.u.......%
000000dc: 00 75 04 95 01 91 03 09 70 15 00 25 64 75 08 95 04 91 02 09  .u......p..%du......
000000f0: 50 66 01 10 55 0e 15 00 26 ff 00 75 08 95 01 91 02 09 a7 15  Pf..U...&..u........
00000104: 00 26 ff 00 75 08 95 01 91 02 65 00 55 00 09 7c 15 00 26 ff  .&..u.....e.U..|..&.
00000118: 00 75 08 95 01 91 02 c0 85 04 05 06 09 20 15 00 26 ff 00 75  .u........... ..&..u
0000012c: 08 95 01 81 02 c0                                            ......
3445511648 1458

```

## Controller and Bluetooth information (for the zero2)

[xpadneo-btmon.txt](https://github.com/atar-axis/xpadneo/files/4908876/xpadneo-btmon.txt)

[xpadneo-dmesg.txt](https://github.com/atar-axis/xpadneo/files/4908877/xpadneo-dmesg.txt)

**Comment by kakra** [comment]
```
[63032.114975] xpadneo: hello there!
[63040.284204] xpadneo 0005:045E:02E0.001D: unknown main item tag 0x0
[63040.284343] input: 8BitDo Zero 2 gamepad as /devices/pci0000:00/0000:00:01.3/0000:01:00.0/usb1/1-2/1-2:1.0/bluetooth/hci0/hci0:256/0005:045E:02E0.001D/input/input57
[63040.284451] xpadneo 0005:045E:02E0.001D: input,hidraw7: BLUETOOTH HID v9.03 Gamepad [8BitDo Zero 2 gamepad] on d4:d2:52:a9:4f:d3
[63492.176904] xpadneo: goodbye!
```

The log seems strange. Did you mix different xpadneo versions? `hello there!` should no longer be present in current versions.

**Comment by terencode** [comment]
Probably bad copy paste, I was trying to know if it's a regression but it doesn't same to be the case.
Let me know if you want me to redo something with latest driver but I can certify it's the same with the latest version.

**Comment by kakra** [comment]
Did you manually make some SDL mappings in the past? E.g., look in `$HOME/.local/share/gamecontrollerdb.local.txt`. Also, do these buttons show up correctly in `evtest`?
`

**Comment by terencode** [comment]
I actually did make one for the SN30 pro+ but not the zero 2 using the controllermap program.
Here are the mappings (this is also what shows up in env):
```
"050000005e040000e002000001000000,8BitDo SN30 Pro+,platform:Linux,a:b0,b:b1,x:b2,y:b3,back:b6,guide:b8,start:b7,leftstick:b9,rightstick:b10,leftshoulder:b4,rightshoulder:b5,dpup:h0.1,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,leftx:a0,lefty:a1,rightx:a3,righty:a4,lefttrigger:a2,righttrigger:a5,"
```
The buttons show up in evtest:
pressing 'y' button: 
```
Event: time 1594576476.189905, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90003
Event: time 1594576476.189905, type 1 (EV_KEY), code 307 (BTN_NORTH), value 1
Event: time 1594576476.189905, -------------- SYN_REPORT ------------
Event: time 1594576476.314848, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90003
Event: time 1594576476.314848, type 1 (EV_KEY), code 307 (BTN_NORTH), value 0
Event: time 1594576476.314848, -------------- SYN_REPORT ------------
```

pressing 'R' button:

```
Event: time 1594576499.643199, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90006
Event: time 1594576499.643199, type 1 (EV_KEY), code 311 (BTN_TR), value 1
Event: time 1594576499.643199, -------------- SYN_REPORT ------------
Event: time 1594576499.718180, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90006
Event: time 1594576499.718180, type 1 (EV_KEY), code 311 (BTN_TR), value 0
Event: time 1594576499.718180, -------------- SYN_REPORT ------------
```

**Comment by terencode** [comment]
I unset the `SDL_GAMECONTROLLERCONFIG` and tried the programs again and same behaviour.

**Comment by kakra** [comment]
Oh, funny... Your controller has X and Y swapped, and also A and B. Did you remap with the controllermap program for this reason? Should be fix this in the driver?

**Comment by kakra** [comment]
```
"050000005e040000e002000001000000
```

This ID will also match your new controller and apply the same remapping to it. That may be the problem. During v0.8 we will add a remapping capability right into the driver in a way that it will detect the exact device. It will even work correctly with two identical models and still see them as different devices with different mappings.

**Comment by terencode** [comment]
> This ID will also match your new controller and apply the same remapping to it

Hmm that's weird I tried it again and while the ID for both the SN30pro+ and the zero2 are the same, they are now different from what they were before (030000005e040000e002000000006800,Xbox One S Controller)

> Your controller has X and Y swapped, and also A and B. Did you remap with the controllermap program for this reason?

I think so. I guess now that you implemented quirks for these devices it makes sense to fix this within the driver?

> Do all 8Bitdo controllers use Nintendo mappings?

I would think so as they are first intended to be used on nintendo games.

**Comment by kakra** [comment]
SDL matches by ID only. The name is just what is presented to the application, you can rename your controller there if you'd want to. The first part `0300` is the bus number. `0500` is for Bluetooth. You probably connected by USB now. Actually, the order of bytes is swapped: `0300` should be read as `0003`, `5e04` should be read as  `045e` (which is the PID for Microsoft).

If you confirm that all buttons work correctly (aside from being swapped) after you removed the mappings, I'll implement a quirk handler for that.

**Comment by terencode** [comment]
I rebooted after removing SDL_GAMECONTROLLERCONFIG and confirmed it with the absence of anything SDL related within `env` but the 'R', 'Y', left and right joystick click (only with SN30 pro+) are still not working in controllermap and antimicrox.

**Comment by kakra** [comment]
Please try if https://github.com/kakra/xpadneo/tree/queue/for-0.9 makes a difference.

For the rest, I need `btmon` logs: Capture the `a1 01` packets and tell me which button you pressed and what logs were generated for it. I'll have to go now, will look at it later.

**Comment by terencode** [comment]
> Please try if https://github.com/kakra/xpadneo/tree/queue/for-0.9 makes a difference. 

No difference unfortunately, previously mentioned buttons not registered. However the 'X' button seems to correspond but the A and B buttons are still swapped.

> For the rest, I need btmon logs: Capture the a1 01 packets and tell me which button you pressed and what logs were generated for it. I'll have to go now, will look at it later.

I'm not sure how to do that. 
Here I used `sudo btmon`, pressed A, B , Y, X, L, R in this order (Y and R being the ones not working in SDL apps) and redirected the output to a file:
[btmon.txt](https://github.com/atar-axis/xpadneo/files/4909259/btmon.txt)

**Comment by kakra** [comment]
Thanks, I've extracted the button data from the packets sent by the controller. It matches exactly with the quirks settings I've applied:
```
ID -- -- -- -- -- -- -- -- -- -- -- -- Buttons
01 00 80 00 80 00 80 00 80 00 00 00 00 00 02 00 - A 0000 0010 = 90002
01 00 80 00 80 00 80 00 80 00 00 00 00 00 00 00 - A 0000 0010
01 00 80 00 80 00 80 00 80 00 00 00 00 00 01 00 - B 0000 0001 = 90001
01 00 80 00 80 00 80 00 80 00 00 00 00 00 00 00 - B 0000 0001
01 00 80 00 80 00 80 00 80 00 00 00 00 00 04 00 - Y 0000 0100 = 90003
01 00 80 00 80 00 80 00 80 00 00 00 00 00 00 00 - Y 0000 0100
01 00 80 00 80 00 80 00 80 00 00 00 00 00 08 00 - X 0000 1000 = 90004
01 00 80 00 80 00 80 00 80 00 00 00 00 00 00 00 - X 0000 1000
01 00 80 00 80 00 80 00 80 00 00 00 00 00 10 00 - L 0001 0000 = 90005 (shoulder)
01 00 80 00 80 00 80 00 80 00 00 00 00 00 00 00 - L 0001 0000
01 00 80 00 80 00 80 00 80 00 00 00 00 00 20 00 - R 0010 0000 = 90006 (shoulder)
01 00 80 00 80 00 80 00 80 00 00 00 00 00 00 00 - R 0010 0000
```

The bits set and reset in the `btmon` packets match exactly with the usage map in the source code.

So what does it tell us? I think the problem is that we fixup the buttons of the reports from Linux to Windows mode but your controller already reports in Windows mode. I need to carry a flag to fix this. Hang on...

**Comment by kakra** [comment]
Do you know the MAC addresses of both your controllers?

**Comment by terencode** [comment]
SN30 pro+ is E4:17:D8:37:9D:61 and zero 2 is E4:17:D8:AA:02:82

**Comment by kakra** [comment]
Please test #229.

**Comment by kakra** [comment]
Please confirm that A, B, X, Y are in the correct order in `jstest`, and that `evtest` shows the correct directions east, south, north, west.

**Comment by kakra** [comment]
Also, please show the dmesg output of this new driver version.

**Comment by terencode** [comment]
I'm not sure how to make sure A, B, X, Y are in correct order in `jstest` but in `evtest` X and Y are swapped (Y says NORTH and X WEST).

In antimicrox, the same buttons still do not work and B and A are swapped.


dmesg output:
```
[11992.109220] xpadneo 0005:045E:02E0.000E: fixing up report size
[11992.109370] xpadneo 0005:045E:02E0.000E: battery detected
[11992.109371] xpadneo 0005:045E:02E0.000E: enabling compliance with Linux Gamepad Specification
[11992.109413] input: 8BitDo Zero 2 gamepad as /devices/pci0000:00/0000:00:01.3/0000:01:00.0/usb1/1-2/1-2:1.0/bluetooth/hci0/hci0:256/0005:045E:02E0.000E/input/input40
[11992.109644] xpadneo 0005:045E:02E0.000E: input,hidraw7: BLUETOOTH HID v9.03 Gamepad [8BitDo Zero 2 gamepad] on d4:d2:52:a9:4f:d3
[11992.109649] xpadneo 0005:045E:02E0.000E: controller quirks: 0x0027
[11992.770288] xpadneo 0005:045E:02E0.000E: 8BitDo Zero 2 gamepad [e4:17:d8:aa:02:82] connected
```

**Comment by kakra** [comment]
`jstest` shows button 0, 1, 2, 3... This should match in order with A, B, X, Y.

**Comment by terencode** [comment]
Ok so with jstest, A=1, B=0, X=3, Y=2

**Comment by kakra** [comment]
Okay, this matches with `evtest` at least and all buttons respond now. Now let's see why it didn't swap A with B, and X with Y... Whatever this is, you should now see all buttons in antimicro and in SDL applications.

**Comment by terencode** [comment]
All buttons were always responding in evtest and jstest, but not in antimicrox, `controllermap`, cemu (wine) and probably more.

**Comment by kakra** [comment]
Could you send another `evtest` output and press all the buttons in order A, B, X, Y, L, R?

**Comment by terencode** [comment]
```
Event: time 1594594606.536562, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90002
Event: time 1594594606.536562, type 1 (EV_KEY), code 305 (BTN_EAST), value 1
Event: time 1594594606.536562, -------------- SYN_REPORT ------------
Event: time 1594594606.687586, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90002
Event: time 1594594606.687586, type 1 (EV_KEY), code 305 (BTN_EAST), value 0
Event: time 1594594606.687586, -------------- SYN_REPORT ------------
Event: time 1594594608.790414, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90001
Event: time 1594594608.790414, type 1 (EV_KEY), code 304 (BTN_SOUTH), value 1
Event: time 1594594608.790414, -------------- SYN_REPORT ------------
Event: time 1594594608.966446, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90001
Event: time 1594594608.966446, type 1 (EV_KEY), code 304 (BTN_SOUTH), value 0
Event: time 1594594608.966446, -------------- SYN_REPORT ------------
Event: time 1594594612.267866, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90004
Event: time 1594594612.267866, type 1 (EV_KEY), code 308 (BTN_WEST), value 1
Event: time 1594594612.267866, -------------- SYN_REPORT ------------
Event: time 1594594612.376455, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90004
Event: time 1594594612.376455, type 1 (EV_KEY), code 308 (BTN_WEST), value 0
Event: time 1594594612.376455, -------------- SYN_REPORT ------------
Event: time 1594594612.755355, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90003
Event: time 1594594612.755355, type 1 (EV_KEY), code 307 (BTN_NORTH), value 1
Event: time 1594594612.755355, -------------- SYN_REPORT ------------
Event: time 1594594612.876366, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90003
Event: time 1594594612.876366, type 1 (EV_KEY), code 307 (BTN_NORTH), value 0
Event: time 1594594612.876366, -------------- SYN_REPORT ------------
Event: time 1594594613.881484, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90005
Event: time 1594594613.881484, type 1 (EV_KEY), code 310 (BTN_TL), value 1
Event: time 1594594613.881484, -------------- SYN_REPORT ------------
Event: time 1594594614.035520, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90005
Event: time 1594594614.035520, type 1 (EV_KEY), code 310 (BTN_TL), value 0
Event: time 1594594614.035520, -------------- SYN_REPORT ------------
Event: time 1594594614.342706, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90006
Event: time 1594594614.342706, type 1 (EV_KEY), code 311 (BTN_TR), value 1
Event: time 1594594614.342706, -------------- SYN_REPORT ------------
Event: time 1594594614.471567, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90006
Event: time 1594594614.471567, type 1 (EV_KEY), code 311 (BTN_TR), value 0
Event: time 1594594614.471567, -------------- SYN_REPORT ------------
```

**Comment by kakra** [comment]
> I double checked and I have nothing related to SDL in `env`.

Please check that there is no file `$HOME/.local/share/gamecontrollerdb.local.txt` which is equivalent to having the SDL environment set. Also, please keep in mind that this file may be taken from `/root` if you are testing as root, and from another directory when running games under some user.

**Comment by terencode** [comment]
I checked both $HOME/.local/share/ and /root/.local/share/ and no `gamecontrollerdb.local.txt` there.

**Comment by kakra** [comment]
I may have found the problem, please test #229 again.

The Linux spec says A = south (mismatch), B = east (mismatch), X = north (match), Y = west (match), so A and B where still swapped.

**Comment by terencode** [comment]
Weird :sweat_smile: unfortunately X and Y are still swapped with evtest.
With jstest: B=0, A=1, X=3, Y=2 (same as before).

**Comment by kakra** [comment]
Could you show the output of `jscal -q`?

**Comment by terencode** [comment]
`jscal -u 8,0,1,2,3,4,5,16,17,10,304,305,307,308,310,311,314,315,317,318 /dev/input/js0`

**Comment by kakra** [comment]
BTW: `evtest` shows cardinal directions, don't rely on that. `evtest` should match the following: A = south, B = east, X = north, Y = west. Games are not using cardinal directions, they go for A, B, X, Y, so you need to correct that in your mind.

**Comment by kakra** [comment]
So if you press A, B, X, Y in that order, `evtest` should show south, east, north, west in that order. Can you confirm?

**Comment by terencode** [comment]
> BTW: evtest shows cardinal directions, don't rely on that

Damn that's really confusing..

> So if you press A, B, X, Y in that order, evtest should show south, east, north, west in that order. Can you confirm?
So, here, when I press A I get EAST; B is SOUTH; Y is NORTH, X is WEST.

**Comment by kakra** [comment]
Yes... Okay, let's start different...
```bash
evtest | sed 's/SOUTH/A/;s/EAST/B/;s/NORTH/X/;s/WEST/Y/'
```

Should reduce headache...

**Comment by terencode** [comment]
With that, A = B, B = 1, Y = X and X = Y.

**Comment by kakra** [comment]
Probably B = A not 1 ;-)

**Comment by terencode** [comment]
Yeah this is messing me up :laughing: I was actually editing it to be easier to parse for the eyes...

**Comment by kakra** [comment]
I shared a lot more hours with this mess, don't ask me... ;-)

With the above sed included, please press A, B, X, Y in that order again and post the output.

**Comment by terencode** [comment]
```
Event: time 1594596847.144665, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90002
Event: time 1594596847.144665, type 1 (EV_KEY), code 305 (BTN_B), value 1
Event: time 1594596847.144665, -------------- SYN_REPORT ------------
Event: time 1594596847.269687, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90002
Event: time 1594596847.269687, type 1 (EV_KEY), code 305 (BTN_B), value 0
Event: time 1594596847.269687, -------------- SYN_REPORT ------------
Event: time 1594596847.801173, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90001
Event: time 1594596847.801173, type 1 (EV_KEY), code 304 (BTN_A), value 1
Event: time 1594596847.801173, -------------- SYN_REPORT ------------
Event: time 1594596847.929770, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90001
Event: time 1594596847.929770, type 1 (EV_KEY), code 304 (BTN_A), value 0
Event: time 1594596847.929770, -------------- SYN_REPORT ------------
Event: time 1594596848.761296, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90004
Event: time 1594596848.761296, type 1 (EV_KEY), code 308 (BTN_Y), value 1
Event: time 1594596848.761296, -------------- SYN_REPORT ------------
Event: time 1594596848.915891, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90004
Event: time 1594596848.915891, type 1 (EV_KEY), code 308 (BTN_Y), value 0
Event: time 1594596848.915891, -------------- SYN_REPORT ------------
Event: time 1594596849.197351, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90003
Event: time 1594596849.197351, type 1 (EV_KEY), code 307 (BTN_X), value 1
Event: time 1594596849.197351, -------------- SYN_REPORT ------------
Event: time 1594596849.338520, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90003
Event: time 1594596849.338520, type 1 (EV_KEY), code 307 (BTN_X), value 0
Event: time 1594596849.338520, -------------- SYN_REPORT ------------
```

**Comment by kakra** [comment]
A is still sending scan code 90002 (which is the original scan code from `btmon` but it should have matched a quirk mapping). I guess my quirk mapping isn't applied. Your quirk code is `0x27` which says "nintendo mapping"... Give me a few minutes.

**Comment by kakra** [comment]
Please try #229 again, reconnect the controller and show the dmesg output...

**Comment by terencode** [comment]
```
[15466.632922] hid-generic 0005:045E:02E0.0019: unknown main item tag 0x0
[15466.633028] input: 8BitDo Zero 2 gamepad as /devices/pci0000:00/0000:00:01.3/0000:01:00.0/usb1/1-2/1-2:1.0/bluetooth/hci0/hci0:256/0005:045E:02E0.0019/input/input54
[15466.633101] hid-generic 0005:045E:02E0.0019: input,hidraw7: BLUETOOTH HID v9.03 Gamepad [8BitDo Zero 2 gamepad] on d4:d2:52:a9:4f:d3
[15466.667641] xpadneo 0005:045E:02E0.0019: fixing up report size
[15466.667761] xpadneo 0005:045E:02E0.0019: mapping hid 0x90001 quirks 0x0000 mask 0x0000 map 0x90001
[15466.667762] xpadneo 0005:045E:02E0.0019: mapping hid 0x90002 quirks 0x0000 mask 0x0000 map 0x90002
[15466.667763] xpadneo 0005:045E:02E0.0019: mapping hid 0x90003 quirks 0x0000 mask 0x0000 map 0x90003
[15466.667763] xpadneo 0005:045E:02E0.0019: mapping hid 0x90004 quirks 0x0000 mask 0x0000 map 0x90004
[15466.667764] xpadneo 0005:045E:02E0.0019: mapping hid 0x90005 quirks 0x0000 mask 0x0000 map 0x90005
[15466.667765] xpadneo 0005:045E:02E0.0019: mapping hid 0x90006 quirks 0x0000 mask 0x0000 map 0x90006
[15466.667765] xpadneo 0005:045E:02E0.0019: mapping hid 0x90007 quirks 0x0000 mask 0x0000 map 0x90007
[15466.667766] xpadneo 0005:045E:02E0.0019: mapping hid 0x90008 quirks 0x0000 mask 0x0000 map 0x90008
[15466.667766] xpadneo 0005:045E:02E0.0019: mapping hid 0x90009 quirks 0x0000 mask 0x0000 map 0x90009
[15466.667767] xpadneo 0005:045E:02E0.0019: mapping hid 0x9000a quirks 0x0000 mask 0x0000 map 0x9000a
[15466.667768] xpadneo 0005:045E:02E0.0019: mapping hid 0x10085 quirks 0x0000 mask 0x0000 map 0x10085
[15466.667768] xpadneo 0005:045E:02E0.0019: battery detected
[15466.667769] xpadneo 0005:045E:02E0.0019: enabling compliance with Linux Gamepad Specification
[15466.667798] input: 8BitDo Zero 2 gamepad as /devices/pci0000:00/0000:00:01.3/0000:01:00.0/usb1/1-2/1-2:1.0/bluetooth/hci0/hci0:256/0005:045E:02E0.0019/input/input55
[15466.667866] xpadneo 0005:045E:02E0.0019: input,hidraw7: BLUETOOTH HID v9.03 Gamepad [8BitDo Zero 2 gamepad] on d4:d2:52:a9:4f:d3
[15466.667876] xpadneo 0005:045E:02E0.0019: controller quirks: 0x00000027
[15467.328285] xpadneo 0005:045E:02E0.0019: 8BitDo Zero 2 gamepad [e4:17:d8:aa:02:82] connected
```

**Comment by kakra** [comment]
Okay, so we have two bugs here... One is: The nintendo quirk is applied too late, otherwise you should see `quirks 0x0020` in the debug log. The second is: The quirk matching is somehow ignored, it's not showing the expected result on my side, too.

**Comment by kakra** [comment]
It's late... There was a brainbug in my head... ;-) Now the bugs in the debug output should be gone and the quirks flag is applied more early. Please pull #229 again and show dmesg.

**Comment by terencode** [comment]
Let's see...

```
[16291.789356] hid-generic 0005:045E:02E0.001A: unknown main item tag 0x0
[16291.789456] input: 8BitDo Zero 2 gamepad as /devices/pci0000:00/0000:00:01.3/0000:01:00.0/usb1/1-2/1-2:1.0/bluetooth/hci0/hci0:256/0005:045E:02E0.001A/input/input56
[16291.789540] hid-generic 0005:045E:02E0.001A: input,hidraw7: BLUETOOTH HID v9.03 Gamepad [8BitDo Zero 2 gamepad] on d4:d2:52:a9:4f:d3
[16291.815039] xpadneo 0005:045E:02E0.001A: fixing up report size
[16291.815155] xpadneo 0005:045E:02E0.001A: mapping hid 0x90001 quirks 0x0000 mask 0x0000 map (1,0x130)
[16291.815157] xpadneo 0005:045E:02E0.001A: mapping hid 0x90002 quirks 0x0000 mask 0x0000 map (1,0x131)
[16291.815158] xpadneo 0005:045E:02E0.001A: mapping hid 0x90003 quirks 0x0000 mask 0x0000 map (1,0x133)
[16291.815159] xpadneo 0005:045E:02E0.001A: mapping hid 0x90004 quirks 0x0000 mask 0x0000 map (1,0x134)
[16291.815160] xpadneo 0005:045E:02E0.001A: mapping hid 0x90005 quirks 0x0000 mask 0x0000 map (1,0x136)
[16291.815160] xpadneo 0005:045E:02E0.001A: mapping hid 0x90006 quirks 0x0000 mask 0x0000 map (1,0x137)
[16291.815161] xpadneo 0005:045E:02E0.001A: mapping hid 0x90007 quirks 0x0000 mask 0x0000 map (1,0x13a)
[16291.815162] xpadneo 0005:045E:02E0.001A: mapping hid 0x90008 quirks 0x0000 mask 0x0000 map (1,0x13b)
[16291.815163] xpadneo 0005:045E:02E0.001A: mapping hid 0x90009 quirks 0x0000 mask 0x0000 map (1,0x13d)
[16291.815164] xpadneo 0005:045E:02E0.001A: mapping hid 0x9000a quirks 0x0000 mask 0x0000 map (1,0x13e)
[16291.815165] xpadneo 0005:045E:02E0.001A: mapping hid 0x10085 quirks 0x0000 mask 0x0000 map (1,0xac)
[16291.815166] xpadneo 0005:045E:02E0.001A: battery detected
[16291.815167] xpadneo 0005:045E:02E0.001A: enabling compliance with Linux Gamepad Specification
[16291.815204] input: 8BitDo Zero 2 gamepad as /devices/pci0000:00/0000:00:01.3/0000:01:00.0/usb1/1-2/1-2:1.0/bluetooth/hci0/hci0:256/0005:045E:02E0.001A/input/input57
[16291.815371] xpadneo 0005:045E:02E0.001A: input,hidraw7: BLUETOOTH HID v9.03 Gamepad [8BitDo Zero 2 gamepad] on d4:d2:52:a9:4f:d3
[16291.815373] xpadneo 0005:045E:02E0.001A: controller quirks: 0x00000027
[16292.475761] xpadneo 0005:045E:02E0.001A: 8BitDo Zero 2 gamepad [e4:17:d8:aa:02:82] connected
```

BTW: Let me know if you prefer to keep going later, it's not a problem ;)

**Comment by terencode** [comment]
When running evtest, Y becomes BTN_C ??

**Comment by kakra** [comment]
Okay, this looks better but clearly the early quirk flag is still missing for you... It should say 0x0020. I'll need to dig into the code a little. Meanwhile, try to load the module with `modprobe hid-xpadneo quirks=e4:17:d8:aa:02:82:0x27` and show dmesg (don't forget to rmmod first).

**Comment by terencode** [comment]
```
[16668.085290] xpadneo 0005:045E:02E0.001B: fixing up report size
[16668.085391] xpadneo 0005:045E:02E0.001B: mapping hid 0x90001 quirks 0x0000 mask 0x0000 map (1,0x130)
[16668.085392] xpadneo 0005:045E:02E0.001B: mapping hid 0x90002 quirks 0x0000 mask 0x0000 map (1,0x131)
[16668.085393] xpadneo 0005:045E:02E0.001B: mapping hid 0x90003 quirks 0x0000 mask 0x0000 map (1,0x133)
[16668.085394] xpadneo 0005:045E:02E0.001B: mapping hid 0x90004 quirks 0x0000 mask 0x0000 map (1,0x134)
[16668.085395] xpadneo 0005:045E:02E0.001B: mapping hid 0x90005 quirks 0x0000 mask 0x0000 map (1,0x136)
[16668.085396] xpadneo 0005:045E:02E0.001B: mapping hid 0x90006 quirks 0x0000 mask 0x0000 map (1,0x137)
[16668.085397] xpadneo 0005:045E:02E0.001B: mapping hid 0x90007 quirks 0x0000 mask 0x0000 map (1,0x13a)
[16668.085398] xpadneo 0005:045E:02E0.001B: mapping hid 0x90008 quirks 0x0000 mask 0x0000 map (1,0x13b)
[16668.085399] xpadneo 0005:045E:02E0.001B: mapping hid 0x90009 quirks 0x0000 mask 0x0000 map (1,0x13d)
[16668.085400] xpadneo 0005:045E:02E0.001B: mapping hid 0x9000a quirks 0x0000 mask 0x0000 map (1,0x13e)
[16668.085401] xpadneo 0005:045E:02E0.001B: mapping hid 0x10085 quirks 0x0000 mask 0x0000 map (1,0xac)
[16668.085402] xpadneo 0005:045E:02E0.001B: battery detected
[16668.085403] xpadneo 0005:045E:02E0.001B: enabling compliance with Linux Gamepad Specification
[16668.085452] input: 8BitDo Zero 2 gamepad as /devices/pci0000:00/0000:00:01.3/0000:01:00.0/usb1/1-2/1-2:1.0/bluetooth/hci0/hci0:256/0005:045E:02E0.001B/input/input59
[16668.085525] xpadneo 0005:045E:02E0.001B: input,hidraw7: BLUETOOTH HID v9.03 Gamepad [8BitDo Zero 2 gamepad] on d4:d2:52:a9:4f:d3
[16668.085528] xpadneo 0005:045E:02E0.001B: quirks override: e4:17:d8:aa:02:82
[16668.085529] xpadneo 0005:045E:02E0.001B: controller quirks: 0x00000027
[16668.746394] xpadneo 0005:045E:02E0.001B: 8BitDo Zero 2 gamepad [e4:17:d8:aa:02:82] connected
```

**Comment by kakra** [comment]
> When running evtest, Y becomes BTN_C ??

According to your dmesg output, there's no way to see `BTN_C` because that would be code `0x132` an that is not mapped anywhere... Strange.

**Comment by terencode** [comment]
> According to your dmesg output, there's no way to see BTN_C because that would be code 0x132 and that is not mapped anywhere... Strange.

And I can't reproduce anymore...

**Comment by terencode** [comment]
Ah, there you go: 
```
Event: time 1594598893.612479, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90003
Event: time 1594598893.612479, type 1 (EV_KEY), code 306 (BTN_C), value 1
Event: time 1594598893.612479, -------------- SYN_REPORT ------------
Event: time 1594598893.837524, type 4 (EV_MSC), code 4 (MSC_SCAN), value 90003
Event: time 1594598893.837524, type 1 (EV_KEY), code 306 (BTN_C), value 0
Event: time 1594598893.837524, -------------- SYN_REPORT ------------
```

No idea how to reproduce it consistently.

**Comment by kakra** [comment]
If this happens, xpadneo is not loaded... 90003 to BTN_C is the standard Linux mapping.

**Comment by terencode** [comment]
You're right, oups... 
Going to bed now, can't delay anymore :D
Hopefuly we'll be done with this mess soon!

**Comment by kakra** [comment]
Okay, I've come to the conclusion that we cannot map by quirk matching because the mapping is done before we can set our quirks. So I'll rewrite the code to swap the bits instead in the report. I don't like that idea too much as this messes with the physical positions games may expect but it should finally fix this.

**Comment by kakra** [comment]
I think we got this, try #229 again.

**Comment by terencode** [comment]
Running `evtest | sed 's/SOUTH/A/;s/EAST/B/;s/NORTH/X/;s/WEST/Y/'` gives me:

- Pressing A gives B
- Pressing B gives A
- Pressing X gives Y
- Pressing Y gives X

controllermap and antimicrox still does not detect Y, R left/right joystick click.

**Comment by kakra** [comment]
Now I'm a bit lost because when I set quirks 0x27, I have my buttons swapped on the Xbox controller (which is has a different layout to yours), and when I set quirks 0x07, I have normal buttons. PS: Actually, my quirks are a little different, it's the "2" that makes the difference.

**Comment by kakra** [comment]
That is, until it doesn't... I tested this by manually setting the quirks within the source code. Putting it on the cmdline doesn't work for some reason. :-(

**Comment by kakra** [comment]
Okay, I think I found it. Your reports are 1 byte shorter than the original Xbox controller reports: only 16 bytes instead of 17, so it wasn't applying the button swap. Should work now, try #229 again and also post dmesg if it didn't work.

**Comment by terencode** [comment]
Yep, now the button swap is working!

Something weird: when I try it in controllermap or antimicrox, my 'Y' button is working now but it shows as 'X' and 'X' button does nothing.
Moreover, the 'L' button is recognized as 'Y' and the 'R' button still does nothing.

**Comment by terencode** [comment]
I installed `sdl2-jstest` and when I run it, it gives me this:

```
GameControllerConfig:
  Name:    'Xbox One S Controller'
  Mapping: '030000005e040000e002000000006800,*,a:b0,b:b1,back:b4,dpdown:b12,dpleft:b13,dpright:b14,dpup:b11,guide:b5,leftshoulder:b9,leftstick:b7,lefttrigger:a4,leftx:a0,lefty:a1,rightshoulder:b10,rightstick:b8,righttrigger:a5,rightx:a2,righty:a3,start:b6,x:b2,y:b3,'
```

**Comment by kakra** [comment]
`a:b0,b:b1,x:b2,y:b3` as it matches the bits we have fixed. But `leftshoulder` and `rightshoulder` is wrong, it should be `b4` and `b5`. You should take care that no SDL mapping exists, neither in env nor as a file. For antimicro, I can only guess that it's also using SDL, or it is reading HID reports directly which it shouldn't do. I don't know how antimicro is implemented but if - after cleanup - antimicro is the only problematic software left, that should be a new bug report.

Does plain `jstest` and `evtest` work correctly now?

**Comment by terencode** [comment]
As I said I don't see any env or file related to that. 
Also it's not just antimicrox, it's `controllermap` and sdl2-jstest too.

Running SDL2 gamepad tool I get this:
```
Searching gamepads...
Found 1 gamepad(s):
"Xbox One S Controller", 030000005e040000e002000000006800 (mapping available)
Environment variable "SDL_GAMECONTROLLERCONFIG" is not defined
Checking if new mappings available from github: https://github.com/gabomdq/SDL_GameControllerDB
```
I also tried lsof when sdl2-jstest is running but no mapping file opened.
It reports it can't find any anyway:
```
GameControllerConfig:
  missing (see 'gamecontrollerdb.txt' or SDL_GAMECONTROLLERCONFIG)
```
So what's going on? 

> Do plain jstest and evtest work correctly now?

Yup reports seem correct now.

**Comment by kakra** [comment]
> "Xbox One S Controller", 030000005e040000e002000000006800 (mapping available)

This line is what alerts me: Why is there a mapping available? Also, the firmware version `00006800` in the end is strange. As far as I see your dmesg shows 9.03, that should make `03090000`. According to http://hg.libsdl.org/SDL/file/default/src/joystick/SDL_gamecontrollerdb.h only two controllers have them, one is a Nintendo Gamecube controller, and one is a PS4 controller, and I think the PS4 controller is probably emulated by 8BitDo and thus shows and original Gamecube firmware version number.

You could try setting `SDL_GAMECONTROLLER_USE_BUTTON_LABELS=1` or `SDL_GAMECONTROLLER_USE_BUTTON_LABELS=0`.

OTOH, you probably just connected the controller via USB - it's not showing the Bluetooth bus number in your ID: `0500`. And then it's not using xpadneo at all because xpadneo does not work with USB connections (except maybe for some obscure edge cases). This probably also explains why `00006800` is shown instead of `03090000`.

From that perspective, I think it's safe to merge #229 and close this: It's properly working in `jstest` and `evtest`, all the rest is in other layers and should be discussed in a separate issue report.

If you've managed to actually bind a `0300` device to xpadneo, that's a different issue. BTW: It would work if such a device is a HID device - original Xbox controllers are not HID via USB, only via Bluetooth. But if your device is HID and you've bound it to xpadneo, that's a new device from this project's view (8BitDo in USB mode), and it's not a topic for this issue. In that case, it's probably also exposing a different mapping which needs a separate fixup. It's probably in dinput mode when connected via USB.

Hint: Bluetooth and USB connection is usually not compatible, the device looks completely different to drivers on either bus.

**Comment by terencode** [comment]
> OTOH, you probably just connected the controller via USB 

Nope. I used both devices using bluetooth.

**Comment by terencode** [comment]
> From that perspective, I think it's safe to merge #229 and close this: It's properly working in jstest and evtest, all the rest is in other layers and should be discussed in a separate issue report.

But at the start this issue was actually about the buttons not being detected, not the fact they were swapped.

> You could try setting SDL_GAMECONTROLLER_USE_BUTTON_LABELS=1 or SDL_GAMECONTROLLER_USE_BUTTON_LABELS=0.

I didn't get this part. Test with what program? It didn't seem to change anything with sdl2-jstest.


- Both devices connected over bluetooth (like I did previously)  have firmware version `030000005e040000e002000000006800`.

**Comment by kakra** [comment]
This is very strange... I probably need one of those controllers physically at hand to fix this. Or we could try some sort of remoting... The controllers should to my best knowledge not show up as ' 0300` when connected over Bluetooth... Except you are using a special dongle just for the controller - but then it's probably technically not Bluetooth.

**Comment by kakra** [comment]
Should not have been closed to to left-over issues.

**Comment by kakra** [comment]
Do you use the 8BitDo Bluetooth dongle by coincidence?

**Comment by kakra** [comment]
> I also tried lsof when sdl2-jstest is running but no mapping file opened.

`lsof` won't show you files that have already been closed again. You'd need to run `strace -f sdl2-jstest` to see the various files it opens and closes. This produces a lot of output and makes the program run slower but you could grep for "controllerdb".

**Comment by terencode** [comment]
> This is very strange... I probably need one of those controllers physically at hand to fix this. Or we could try some sort of remoting...

To set things up you can contact me on Telegram (@terencode),Matrix or Discord (Terence#2723) .

> You could run with SDL_GAMECONTROLLER_... program-to-run to try if it makes a difference. That sets the environment for program-to-run just for this program while it is running.

Yep that's what I tried but no difference unfortunately.

> Do you use the 8BitDo Bluetooth dongle by coincidence?

Nope (using intel Wireless-AC 9260).

> lsof won't show you files that have already been closed again. You'd need to run strace -f sdl2-jstest to see the various files it opens and closes. 

Thanks tried it but nothing stands out except confirmation it is not finding any gamecontrollerdb file:  `openat(AT_FDCWD, "gamecontrollerdb.txt", O_RDONLY) = -1 ENOENT (No such file or directory)`

**Comment by kakra** [comment]
> Nope (using intel Wireless-AC 9260).

I had some problems with this chipset lately - but on Windows, and for Wifi... It needed a driver update to work correctly. But whatever that problem is and where it is located - it should not affect this specific problems. But maybe worth keeping that in mind.

> > You'd need to run strace -f sdl2-jstest to see the various files it opens and closes.
> 
> Thanks tried it but nothing stands out except confirmation it is not finding any gamecontrollerdb file

Let's see if strace can tell us which event or other files from `/dev/` it opens. Do you think you could trace that?

**Comment by kakra** [comment]
> To set things up you can contact me on ... or Discord (Terence#2723) .

https://discord.gg/AHqMBd

**Comment by terencode** [comment]
> Let's see if strace can tell us which event or other files from /dev/ it opens. Do you think you could trace that?

here you go: [sdl2-jstest-strace.txt](https://github.com/atar-axis/xpadneo/files/4941496/sdl2-jstest-strace.txt)

**Comment by kakra** [comment]
An strace of an SDL application would be useful, including `udevadm monitor` (start this before running strace on a second terminal) from the same testing session. Let's see how SDL matches up with the devices detected by udev. SDL should clearly not see bus 3 when the device is connected over Bluetooth. Or is another controller connected at the same time?

**Comment by terencode** [comment]
>  Or is another controller connected at the same time?

Only one controller is connected (zero 2).

[udevadm.txt](https://github.com/atar-axis/xpadneo/files/4941534/udevadm.txt)

[controllermap-strace.txt](https://github.com/atar-axis/xpadneo/files/4941535/controllermap-strace.txt)

**Comment by kakra** [comment]
Hmm, I wonder if that's because it's opening `js0` at the end. It's initially clearly opening the same event device as listet in udevadm, tho, and udevadm confirms it's seen on the Bluetooth bus. I'll compare with my own tests to see if it's doing the same here.

**Comment by kakra** [comment]
What does it say if you just run `controllermap` without arguments?

**Comment by terencode** [comment]
```
INFO: There are 1 joysticks attached
INFO: Joystick 0: Xbox One S Controller
INFO:        axes: 6
INFO:       balls: 0
INFO:        hats: 0
INFO:     buttons: 19
INFO: instance id: 0
INFO:        guid: 030000005e040000e002000000006800
INFO:     VID/PID: 0x045e/0x02e0
INFO: 
```

**Comment by kakra** [comment]
I'm not even sure how that could come from xpadneo because xpadneo exposes 11 buttons at most. Are you sure that this controller is handled by xpadneo? Is there something sitting in between that grabs the device from xpadneo and exposes it differently? Maybe a user space daemon, i.e. the Steam controller daemon that's coming with Steam? You could try turning that off in the Steam client.

See if anything has the device open: `sudo lsof | grep "js0|event20"` (and replace with the names your controller uses).

**Comment by terencode** [comment]
> Are you sure that this controller is handled by xpadneo? 

I would think so because 1) I set it to be in X input mode which emulates an Xbox 360 (https://support.8bitdo.com/faq/sn30-pro.html) and 2) xpadneo picks it up as shown in the different dmseg I posted: `kernel: input: 8BitDo Zero 2 gamepad as /devices/pci0000:00/0000:00:01.3/0000:01:00.0/usb1/1-2/1-2:1.0/bluetooth/hci0/hci0:256/0005:045E:02E0.000C/input/input34
`

>  Maybe a user space daemon, i.e. the Steam controller daemon that's coming with Steam? You could try turning that off in the Steam client.

I had already thought it could conflict but it shouldn't be the case due to the fact steam isn't running.

> See if anything has the device open: sudo lsof | grep "js0|event20" (and replace with the names your controller uses).

I used `grep -E 'js0|event26'` on the output of `sudo lsof` and I only see chromium using it (even when controllermap is running):  [lsof-js.txt](https://github.com/atar-axis/xpadneo/files/4942117/lsof-js.txt)


BTW: I joined the discord server.

**Comment by kakra** [comment]
> `0005:045E:02E0.`

This is actually used by SDL to create the ID number, it would make `050000005e040000e0020000........` from it, but the dots are filled from the version field of the input device (usually the firmware version), in `dmesg` that's just a unique number built from a counter so that two identical models would have two different IDs in the kernel.

By looking at the SDL source code, it copies that straight from whatever udev provides. But then it does some magic: It looks if there's a default or xinput mapping available in the controller DB, and if it then cannot find the device, it tries to fall back to one of the two default mappings (it uses the xinput mapping if it finds a hint for Xbox controllers in the device name). But neither "xinput" nor "default" have the button map you're seeing. Also, your GUID is not in the SDL upstream controller DB.

Something remotely related only exists in the MacOS version of SDL (`030000005e040000e0020000...`) but that actually defines an xpadneo compatible mapping. And SDL doesn't match by device name as far as I can see, so it should not pick one of the 8BitDo mappings. So it seems your mapping entry is completely bogus and comes from somewhere else. This is also indicated by the name string which, when auto-detected, should say "8BitDo Zero 2 gamepad". Something is messed up in your system, something replaces the device. Or there's some obscure SDL environment variable set. Or something intercepts opening the device and emulates it as something else (which could explain the bus number, the Steam UI may do this but only while it is running: `pgrep -a steam`).

Could you run `locate controllerdb` or `find / -type f -name "*controllerdb*"`?

Which SDL version are you running?

Did you compile and install SDL yourself?

Could you run `lddtree $(type -p controllermap)`?

Does you environment set `LD_PRELOAD` or `LD_LIBRARY_PATH`?

In https://github.com/atar-axis/xpadneo/issues/228#issuecomment-657256999 you wrote that the controller ID changed, and I'm guessing that means from `0500...` to `0300...`. What happened to your system? What did you do?

**Comment by kakra** [comment]
I think this may be antimicro... It is probably implemented as a uinput client: It needs to grab the controller and expose it as a new device. But then it should be on the virtual bus (bus 6). I wasn't able to reproduce such a behavior, the only device I was seeing was a virtual mouse. Maybe post `./.config/antimicro/antimicro_settings.ini`.

**Comment by terencode** [comment]
> Which SDL version are you running? Did you compile and install SDL yourself?  Does your distribution patch SDL in a strange way? Or doesn't link to udev? In #228 (comment) you wrote that the controller ID changed, and I'm guessing that means from 0500... to 0300.... What happened to your system? What did you do?

I didn't do much besides upgrading the system regularly, then noticing my gamepad had a different behaviour.
I'm using the one from the arch repo, SDL 2.0.12. 

I found something very interesting though. Look at the the diff of the last packaged version: https://git.archlinux.org/svntogit/packages.git/commit/trunk?h=packages/sdl2&id=b77621a8fbe14a4e06ba640e2b687e8d6092d526

I reverted this and it's working as before!

With the revert, the guid is `030000005e040000e002000000006800` and all buttons are working correctly.
Without, it's `050000005e040000e002000003090000` and the previously mentioned buttons are not working.

Where do we go from here?

I still answered your other questions just in case:

> Could you run locate controllerdb or find / -type f -name "*controllerdb*"?

I already tried that and many times, nothing relevant shows up (not used). [locate.txt](https://github.com/atar-axis/xpadneo/files/4946924/locate.txt)

> Could you run lddtree $(type -p controllermap)?

```
lddtree: warning: controllermap: did not match any paths
lddtree: warning: is: did not match any paths
/usr/bin/controllermap (interpreter => /lib64/ld-linux-x86-64.so.2)
    libSDL2-2.0.so.0 => /usr/lib/libSDL2-2.0.so.0
        libm.so.6 => /usr/lib/libm.so.6
        libdl.so.2 => /usr/lib/libdl.so.2
        libpthread.so.0 => /usr/lib/libpthread.so.0
    libc.so.6 => /usr/lib/libc.so.6
```

> Does you environment set LD_PRELOAD or LD_LIBRARY_PATH?

> Did you accidentally or on purpose modify a system-installed file of SDL?

No.

**Comment by terencode** [comment]
> In udevadm monitor you should see the hidraw device name, you could chmod 0000 and see if SDL still behaves this way.

If I chmod 0000 the hidraw device SDL behaves correctly.

**Comment by kakra** [comment]
So maybe I should revert that patch or we find some other way to exclude that from special handling in SDL... I'll have to dive deep into the patch first.

**Comment by kakra** [comment]
Please try setting `SDL_JOYSTICK_HIDAPI=0` for starting SDL applications.

SDL 2.0.12 introduced a patch that talks to HID devices directly, bypassing drivers. It also doesn't pass the bus number, so it appears on bus 3 in SDL. This patch may be useful but it is quite bad for several reasons: It bypasses fixes that are implemented in drivers, such as that Xbox controllers need to throttle the rumble packets. Also, our fixes for buttons become messed up, the implementation cannot know that we fix the buttons in the HID reports. With this feature, it will not be possible to use some of the additional features we are going to add. IOW, it duplicates a lot of work and bugfixes: SDL would now have to solve the same problems we solved in xpadneo.

The new feature should be off by default, or only use the new implementation as a fallback if the device isn't supported by an OS driver. OTOH, SDL implements a user-space driver for any device regardless of the kernel driver support which isn't too bad. This is similar to how xow implements a user-space driver just that SDL bypasses the Linux HID layer that way.

I'm still on libsdl-2.0.10, let's see what we do about this when I'm upgraded, too. Maybe we need to revert the patch that allows direct raw HID access.

I'm not sure how this is meant to work correctly when SDL doesn't take the bus ID of the device into account because clearly Xbox controllers behave differently based on the bus number. Also, it should query the driver name and make that part of the controller mappings. And it should query and match the number of buttons it gets from the HID descriptor because only then it can know the button mappings to use.

If you disable the HID API of SDL, your controller should show up on bus 5 again and work correctly. Could you test?

**Comment by terencode** [comment]
> Please try setting SDL_JOYSTICK_HIDAPI=0 for starting SDL applications.

Working. I now added it to my env :)

> If you disable the HID API of SDL, your controller should show up on bus 5 again and work correctly. Could you test?

If you mean when using ` SDL_JOYSTICK_HIDAPI=0`, I think it does show up on bus 5: `050000005e040000e002000003090000`. Without, it's `030000005e040000e002000000006800`.

It has the same effect as reverting to the unpatched SDL.

**Comment by kakra** [comment]
The unpatched SDL has a bug that prevented HID API from fully working with some versions of libusb, I believe.

So I think we'll just document the issue with SDL 2.0.12 and can then finally close this?

**Comment by terencode** [comment]
Alright. 
Yes, after ~~91~~ 92 messages we are good to close this.
Thank you very much for all your time spent diagnosing this :)

**Comment by kakra** [comment]
I'll reopen this to close it with a documentation update...

**Comment by terencode** [comment]
Oups, didn't get your comment about the documentation right, sorry.
